core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -326,7 +326,7 @@ class ControllerBrokerRequestBatch(controller: KafkaController, stateChangeLogge
                                        leaderIsrAndControllerEpoch: LeaderIsrAndControllerEpoch,
                                        replicas: Seq[Int], isNew: Boolean) {
 
-    brokerIds.filter(_ >= 0).foreach { brokerId =>
+    brokerIds.filter(b => b >= 0 && controllerContext.liveOrShuttingDownBrokerIds.contains(b)).foreach { brokerId =>
```

Comment:
After introducing broker epoch, we need to fill in the broker epoch we want to use for the control request we send out. Only brokers in `controllerContext.liveOrShuttingDownBrokerIds` will have a entry in the `controllerContext.brokerEpochsCache`.

The benefit of this change is to ensure that we can always get back the broker epoch from the cache in controller context when constructing the control requests. We can instead add extra logic in `controllerChannelManager.sendRequestsToBrokers` to check the existence of the broker epoch when constructing the request but I think it is cleaner and easier to reason about the code with this change because we will not send out requests to brokers that are not in `controllerContext.liveOrShuttingDownBrokerIds` anyway. 

It is true that the broker can become offline and this will cause RequestSendThread to fail to send out the request if controller doesn't process the broker change event before RequestSendThread sending out the request to the offline broker. This change will not affect this behavior and does not aim to solve this race condition. This change only acts as a pre-filter to ensure we can always construct the control request with broker epoch. Whether we can actually send out the request is a separate issue.

Reply:
Sounds good. I am not sure when brokerId in the `brokerIds` will be negative. But since we don't expect any brokerId in `liveOrShuttingDownBrokerIds` to be negative, can we just do `brokerIds.filter(controllerContext.liveOrShuttingDownBrokerIds.contains)` to simplify the code here?

Reply:
That is a good point. I think we can simplify it.

Reply:
Done.